internal/state/change_processor.go

```diff
@@ -71,77 +78,86 @@ func (c *ChangeProcessorImpl) CaptureUpsertChange(obj client.Object) {
 	c.lock.Lock()
 	defer c.lock.Unlock()
 
-	resourceChanged := true
+	c.relationshipCapturer.Capture(obj)
 
 	switch o := obj.(type) {
 	case *v1beta1.GatewayClass:
-		if o.Name != c.cfg.GatewayClassName {
-			panic(fmt.Errorf("gatewayclass resource must be %s, got %s", c.cfg.GatewayClassName, o.Name))
-		}
-		// if the resource spec hasn't changed (its generation is the same), ignore the upsert
-		if c.store.gc != nil && c.store.gc.Generation == o.Generation {
-			resourceChanged = false
-		}
-		c.store.gc = o
+		c.store.captureGatewayClassChange(o, c.cfg.GatewayClassName)
 	case *v1beta1.Gateway:
-		// if the resource spec hasn't changed (its generation is the same), ignore the upsert
-		prev, exist := c.store.gateways[getNamespacedName(obj)]
-		if exist && o.Generation == prev.Generation {
-			resourceChanged = false
-		}
-		c.store.gateways[getNamespacedName(obj)] = o
+		c.store.captureGatewayChange(o)
 	case *v1beta1.HTTPRoute:
-		// if the resource spec hasn't changed (its generation is the same), ignore the upsert
-		prev, exist := c.store.httpRoutes[getNamespacedName(obj)]
-		if exist && o.Generation == prev.Generation {
-			resourceChanged = false
-		}
-		c.store.httpRoutes[getNamespacedName(obj)] = o
+		c.store.captureHTTPRouteChange(o)
+	case *v1.Service:
+		c.store.captureServiceChange(o)
+	case *discoveryV1.EndpointSlice:
+		break
 	default:
 		panic(fmt.Errorf("ChangeProcessor doesn't support %T", obj))
 	}
 
-	c.storeChanged = c.storeChanged || resourceChanged
+	c.store.changed = c.store.changed || c.relationshipCapturer.Exists(obj, client.ObjectKeyFromObject(obj))
```

Comment:
I can introduce a new boolean for the ChangeProcessor. Still, I don't think we can entirely remove its dependency on the store because we need a way to reset the `store.changed` variable after a Process event.

Currently, we do this in `ChangeProcessor.Process()`: https://github.com/nginxinc/nginx-kubernetes-gateway/blob/3c90c2296b5c1caf600f7e6a5ff8b766cd9a0a2c/internal/state/change_processor.go#L136.

Reply:
> c.storeChanged = false 

That looks good to me. In that case, we just reseting the store. However, `c.store.changed = c.store.changed || c.relationshipCapturer.Exists(obj, client.ObjectKeyFromObject(obj))`, we're adding additional meaning to `c.store.changed`, so that it also reflects changed relationship between objects, even if they are not present in the store.